torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -1159,42 +1191,64 @@ class ShapePropagator {
+              node, /*num_reduce_dim=*/*maybe_keepdim?0:1, /*integer_upcast=*/true, opt_dtype);
+        }};
+
+    // Requirements:
+    //   dims           : preserved
+    //   scalar type    : preserved unless specified.
+    //   device         : preserved
+    //   tensor inputs  : 1
+    //   tensor outputs : 1
+    // Additionally:
+    //   - First input should be the only tensor input
+    static const register_formula_for dim_reduce_ops_dtype{
+        {
+          "aten::cumprod(Tensor self, int dim, *, int? dtype) -> Tensor",
+          "aten::cumsum(Tensor self, int dim, *, int? dtype) -> Tensor",
+          "aten::log_softmax(Tensor self, int dim, *, int? dtype) -> Tensor"
         },
         [](Node* node) -> type_vec_t {
-          return multidim_reduce_with_postprocess(
-              node, /*num_reduce_dim=*/1, /*integer_upcast=*/true);
+          at::optional<IValue> opt_dtype = node->get(attr::dtype);
+          return reduce_op_handler(
+              node, /*num_reduce_dim=*/0, /*integer_upcast=*/true, opt_dtype);
         }};
 
     // Requirements:
     //   dims           : preserved if keepdim == false, dim->size() smaller
-    //   otherwise scalar type    : preserved device         : preserved tensor
-    //   inputs  : 1 tensor outputs : 1
+    //   otherwise
+    //   scalar type    : preserved
+    //   device         : preserved
+    //   tensor inputs  : 1
+    //   tensor outputs : 1
     // Additionally:
     //   - First input should be the only tensor input
     //   - has a bool keepdim argument
     static const register_formula_for multidim_reduce_ops{
         {
             "aten::logsumexp(Tensor self, int[] dim, bool keepdim) -> Tensor",
-            "aten::mean(Tensor self, int[] dim, bool keepdim) -> Tensor",
             "aten::norm(Tensor self, Scalar? p, int[] dim, bool keepdim) -> Tensor",
             "aten::std(Tensor self, int[] dim, bool unbiased, bool keepdim) -> Tensor",
             "aten::var(Tensor self, int[] dim, bool unbiased, bool keepdim) -> Tensor",
             "aten::max_values(Tensor self, int[] dim, bool keepdim) -> Tensor",
             "aten::min_values(Tensor self, int[] dim, bool keepdim) -> Tensor",
         },
         [](Node* node) -> type_vec_t {
-          if (auto dim = node->get<std::vector<int64_t>>(attr::dim)) {
-            return multidim_reduce_with_postprocess(
-                node,
-                /*num_reduced_dim=*/dim->size(),
-                /*upcast_integer=*/false);
-          }
-          return {};
+          auto dims = node->namedInput(attr::dim)->node()->inputs().size();
+          return multidim_reduce_with_keepdim(
+              node,
+              /*num_reduced_dim=*/dims,
+              /*upcast_integer=*/false);
         }};
 
     // Requirements:
-    //   dims           : preserved if keepdim == false, 1 smaller otherwise
+    //   dims           : preserved if keepdim == false, dim.size() smaller otherwise
```

Comment:
is this equivalent to saying `preserved if keepdim == false, 0 otherwise`

Reply:
no because dims comes from the `dim` parameter, not from `self`.  E.g., 3-dimension tensor reduced on two dims gives 1-dim tensor:
```
>>> torch.ones([3,3,3]).sum([0,1])
tensor([9., 9., 9.])
```